### PR TITLE
Refactor LibriSpeech Conformer RNN-T recipe

### PR DESCRIPTION
**Pull request by hwangjeff**
Modifies the example LibriSpeech Conformer RNN-T recipe as follows:
- Moves data loading and transforms logic from lightning module to data module (improves generalizability and reusability of lightning module and data module).
- Moves transforms logic from dataloader collator function to dataset (resolves dataloader multiprocessing issues on certain platforms).
- Replaces lambda functions with `partial` equivalents (resolves pickling issues in certain runtime environments).
- Modifies training script to allow for specifying path model checkpoint to restart training from.

### CODE REVIEW

File: examples/asr/librispeech_conformer_rnnt/data_module.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import os
+import random
+
+import torch
+import torchaudio
+from pytorch_lightning import LightningDataModule, seed_everything
+
+
+seed_everything(1)
```
Comment (mthrok) [inline examples/asr/librispeech_conformer_rnnt/data_module.py:9]: Is this the right place to seed? I would imagine this happen once (and only once) at the very beginning of the CLI entry point.

**Review by mthrok — APPROVED** [review]
Stamp

File: examples/asr/librispeech_conformer_rnnt/data_module.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+
+def _batch_by_token_count(idx_target_lengths, token_limit, sample_limit=None):
```
Comment (xiaohui-zhang) [inline examples/asr/librispeech_conformer_rnnt/data_module.py:12]: Can you change sample_limit to batch_size and token_limit to max_tokens per our previous discussion? It's ok if you want to do that in another PR.

File: examples/asr/librispeech_conformer_rnnt/data_module.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                )
+                for dataset, lengths in zip(datasets, self.val_dataset_lengths)
+            ]
+        )
+        dataset = TransformDataset(dataset, self.val_transform)
+        dataloader = torch.utils.data.DataLoader(dataset, batch_size=None, num_workers=self.num_workers)
+        return dataloader
+
+    def test_dataloader(self):
+        dataset = torchaudio.datasets.LIBRISPEECH(self.librispeech_path, url="test-clean")
```
Comment (xiaohui-zhang) [inline examples/asr/librispeech_conformer_rnnt/data_module.py:191]: better make the eval split customizable as we discussed. again I can do that in a later PR as well.

File: examples/asr/librispeech_conformer_rnnt/data_module.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+class LibriSpeechDataModule(LightningDataModule):
+    def __init__(
+        self,
+        *,
+        librispeech_path,
+        train_transform,
+        val_transform,
+        test_transform,
+        max_token_limit=700,
```
Comment (xiaohui-zhang) [inline examples/asr/librispeech_conformer_rnnt/data_module.py:116]: max_tokens should be fine. "max" duplicates with "limit"

File: examples/asr/librispeech_conformer_rnnt/data_module.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        sample_limit=2,
```
Comment (xiaohui-zhang) [inline examples/asr/librispeech_conformer_rnnt/data_module.py:117]: batch_size

File: examples/asr/librispeech_conformer_rnnt/train.py [hunk 6]
```diff
@@ -42,12 +32,9 @@ def run_train(args):
...
 
-    model = ConformerRNNTModule(
-        librispeech_path=str(args.librispeech_path),
-        sp_model_path=str(args.sp_model_path),
-        global_stats_path=str(args.global_stats_path),
-    )
-    trainer.fit(model)
+    model = ConformerRNNTModule(str(args.sp_model_path))
+    data_module = get_data_module(str(args.librispeech_path), str(args.global_stats_path), str(args.sp_model_path))
+    trainer.fit(model, data_module)
```
[the remark this answers is not in the document]
Comment (xiaohui-zhang) [reply examples/asr/librispeech_conformer_rnnt/train.py]: maybe --resume-from-checkpoint
